agent/lib/kontena/workers/service_pod_worker.rb

```diff
@@ -62,17 +67,71 @@ def on_container_event(topic, e)
       if attrs['io.kontena.service.id'] == @service_pod.service_id &&
           attrs['io.kontena.service.instance_number'].to_i == @service_pod.instance_number
         @container_state_changed = true
+        handle_restart_on_die if e.status == 'die'.freeze
+      end
+    end
+
+    # Handles events when container has died
+    def handle_restart_on_die
+      unless @service_pod.running?
+        @restart_backoff_timer.cancel if @restart_backoff_timer
+        return
       end
+
+      return if apply_in_progress?
+
+      # cancel counters because we are called again
+      cancel_restart_timers
+
+      # backoff restarts
+      backoff = @restarts ** 2
+      backoff = max_restart_backoff if backoff > max_restart_backoff
+      if backoff == 0
+        info "restarting #{@service_pod.name_for_humans} because it has stopped"
+      else
+        info "restarting #{@service_pod.name_for_humans} because it has stopped (delay: #{backoff}s)"
+      end
+      @restart_backoff_timer = after(backoff) {
+        apply_restart if @container_state_changed
+      }
+      @restarts += 1
+    end
+
+    def max_restart_backoff
+      Kontena::Workers::ServicePodManager::LOOP_INTERVAL
+    end
+
+    def cancel_restart_timers
+      @restart_counter_reset_timer.cancel if @restart_counter_reset_timer
+      @restart_backoff_timer.cancel if @restart_backoff_timer
+    end
+
+    def apply_restart
+      apply
+      # reset restart counter if instance stays up 10s
+      @restart_counter_reset_timer = after(10) {
+        debug "#{@service_pod.name_for_humans} stayed up 10s, resetting restart backoff counter"
+        @restarts = 0
+      }
+    end
+
+    def restarting?
+      @restarts > 0
     end
 
     def destroy
       @service_pod.mark_as_terminated
       apply
     end
 
+    def apply_in_progress?
```

Comment:
Yes... a `kontena service deploy --force` will trigger a total of three `apply`s now, because all of the container events from the initial apply queue up and get processed:

```
D, [2017-08-21T11:31:45.207297 #1] DEBUG -- Kontena::RpcServer: rpc request: Kontena::Rpc::ServicePodsApi#notify_update [[]]
D, [2017-08-21T11:31:45.557896 #1] DEBUG -- Kontena::Workers::ServicePodWorker: state of redis-test-1: running
I, [2017-08-21T11:31:45.596364 #1]  INFO -- Kontena::Workers::ServicePodWorker: re-creating redis-test-1
I, [2017-08-21T11:31:45.597022 #1]  INFO -- Kontena::ServicePods::Creator: creating service: redis-test-1
I, [2017-08-21T11:31:47.963434 #1]  INFO -- Kontena::ServicePods::Creator: removing previous version of service: redis-test-1
D, [2017-08-21T11:31:48.488478 #1] DEBUG -- Kontena::ServicePods::Creator: creating container: redis-test-1
D, [2017-08-21T11:31:48.630464 #1] DEBUG -- Kontena::ServicePods::Creator: container created: redis-test-1
I, [2017-08-21T11:31:49.001441 #1]  INFO -- Kontena::ServicePods::Creator: service started: redis-test-1
D, [2017-08-21T11:31:49.115770 #1] DEBUG -- Kontena::Workers::ServicePodWorker: on_container_event status=kill => container_state_changed!
D, [2017-08-21T11:31:49.116685 #1] DEBUG -- Kontena::Workers::ServicePodWorker: on_container_event status=die => container_state_changed!
D, [2017-08-21T11:31:49.118750 #1] DEBUG -- Kontena::Workers::ServicePodWorker: handle_restart_on_die: pod running?=true apply_in_progress?=false restarts=0
I, [2017-08-21T11:31:49.118931 #1]  INFO -- Kontena::Workers::ServicePodWorker: restarting redis-test-1 because it has stopped
D, [2017-08-21T11:31:49.119201 #1] DEBUG -- Kontena::Workers::ServicePodWorker: handle_restart_on_die => apply_restart if true
D, [2017-08-21T11:31:49.119442 #1] DEBUG -- Kontena::Workers::ServicePodWorker: state of redis-test-1: running
D, [2017-08-21T11:31:49.264550 #1] DEBUG -- Kontena::Workers::ServicePodWorker: state is in-sync: running
D, [2017-08-21T11:31:49.304540 #1] DEBUG -- Kontena::Workers::ServicePodWorker: on_container_event status=stop => container_state_changed!
D, [2017-08-21T11:31:49.304910 #1] DEBUG -- Kontena::Workers::ServicePodWorker: on_container_event status=destroy => container_state_changed!
D, [2017-08-21T11:31:49.305751 #1] DEBUG -- Kontena::Workers::ServicePodWorker: on_container_event status=create => container_state_changed!
D, [2017-08-21T11:31:49.305928 #1] DEBUG -- Kontena::Workers::ServicePodWorker: on_container_event status=start => container_state_changed!
D, [2017-08-21T11:31:59.304620 #1] DEBUG -- Kontena::Workers::ServicePodWorker: redis-test-1 stayed up 10s, resetting restart backoff counter
D, [2017-08-21T11:32:01.854993 #1] DEBUG -- Kontena::Workers::ServicePodWorker: state of redis-test-1: running
D, [2017-08-21T11:32:01.899455 #1] DEBUG -- Kontena::Workers::ServicePodWorker: state is in-sync: running
```